### PR TITLE
Implement OverlapController

### PR DESCRIPTION
**Pull request by pinguinson**
Added OverlapController (with tests), cleaned up code.

### CODE REVIEW

File: genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import gq.core.data.GQModule
+import gq.core.data.Species
+import gq.core.data.intersectWithSorted
+import gq.core.genes.GeneFormat
+import gq.rest.GQDataRepository
+import gq.rest.exceptions.BadRequestException
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+
+data class OverlapResponse(val identifiedGeneFormat: String,
```
Comment (smolcoder) [inline genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt]: We don't need information about neither `identifiedGeneFormat` nor `geneConversionMap` in this request. Can you please remove it? It's fine to leave just `overlapGenes`.

File: genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                           val geneConversionMap: Map<String, Long?>,
+                           val overlapGenes: LongArray)
+
+@Service
+open class OverlapService @Autowired constructor(
+        private val gqDataRepository: GQDataRepository) {
+
+    open fun convertGenes(rawGenes: List<String>,
```
Comment (smolcoder) [inline genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt]: Duplicated code from [GeneSetEnrichmentService](https://github.com/ctlab/genequery-kotlin/blob/master/genequery-rest/src/main/kotlin/gq/rest/services/GeneSetEnrichmentService.kt#L23) class. This code should be reused.

File: genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            val currentGeneFormat = GeneFormat.guess(rawGenes)
+            return Pair(
+                    currentGeneFormat,
+                    gqDataRepository.smartConverter.toEntrez(rawGenes, currentGeneFormat, speciesFrom, speciesTo))
+        } catch (ex: IllegalArgumentException) {
+            throw BadRequestException(ex)
+        }
+    }
+
+    open fun getModule(moduleName: String) : GQModule {
```
Comment (smolcoder) [inline genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt]: Do we need this function to be `open`?

File: genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    open fun findOverlap(rawGenes: List<String>,
+                         speciesFrom: Species,
+                         speciesTo: Species = speciesFrom,
+                         moduleName: String): OverlapResponse {
+        val (identifiedGeneFormat, conversionMap) = convertGenes(rawGenes, speciesFrom, speciesTo)
+        val module = getModule(moduleName)
+        val entrezIds = conversionMap.values.filterNotNull().toLongArray().sortedArray()
+        val overlapGenes = entrezIds intersectWithSorted module.sortedEntrezIds
+        return OverlapResponse(identifiedGeneFormat.formatName, conversionMap, overlapGenes)
```
Comment (smolcoder) [inline genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt]: As far as I remember we are supposed to return genes in `symbol` format but here is `entrez`. Please, add conversion to `symbol` before returning the result (and, seemingly, modify `OverlapResponse`).

File: genequery-rest/src/test/kotlin/gq/rest/api/GeneSetEnrichmentControllerTest.kt [hunk 5]
```diff
@@ -27,16 +24,6 @@ import org.springframework.web.context.WebApplicationContext
 import java.nio.charset.Charset
 
 
-class PatchEnvPropsAppCtxInitializer : ApplicationContextInitializer<ConfigurableApplicationContext> {
```
Comment (smolcoder) [inline genequery-rest/src/test/kotlin/gq/rest/api/GeneSetEnrichmentControllerTest.kt]: Make sure you're up-to-date with current version of the code base. This class must exist.

File: genequery-rest/src/test/kotlin/gq/rest/api/OverlapControllerTest.kt [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                .andExpect(jsonPath("$.success", equalTo(false)))
+                .andExpect(jsonPath("$.result", nullValue()))
+                .andExpect(jsonPath("$.errors", hasSize<Int>(4)))
+                .andExpect(jsonPath("$.errors", hasItem(containsString("moduleName"))))
+                .andExpect(jsonPath("$.errors", hasItem(containsString("hss"))))
+                .andExpect(jsonPath("$.errors", hasItem(containsString("speciesFrom"))))
+                .andExpect(jsonPath("$.errors", hasItem(containsString("genes"))))
+    }
+
+    private fun makeRequest(form: OverlapController.OverlapRequestForm): ResultActions {
```
Comment (smolcoder) [inline genequery-rest/src/test/kotlin/gq/rest/api/OverlapControllerTest.kt]: Method `makeRequest` and `json` duplicate code from [GeneSetEnrichmentControllerTest](https://github.com/ctlab/genequery-kotlin/blob/master/genequery-rest/src/test/kotlin/gq/rest/api/GeneSetEnrichmentControllerTest.kt). Could you extract it to somewhere else (e.g. superclass)?

File: genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt (hunk 1 above)
Comment (pinguinson) [inline genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt]: Fixed, but not sure if we should remove `OverlapResponse` class or not.

File: genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt (hunk 2 above)
Comment (pinguinson) [inline genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt]: Fixed.

File: genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt (hunk 3 above)
Comment (pinguinson) [inline genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt]: Don't think so, fixed.

File: genequery-rest/src/test/kotlin/gq/rest/api/OverlapControllerTest.kt (hunk 6 above)
Comment (pinguinson) [inline genequery-rest/src/test/kotlin/gq/rest/api/OverlapControllerTest.kt]: Done.

File: .idea/.name [hunk 7]
```diff
@@ -1 +1 @@
-genequery-kotlin
```
Comment (smolcoder) [inline .idea/.name]: wat

File: .idea/compiler.xml [hunk 8]
```diff
@@ -17,5 +17,13 @@
         <processorPath useClasspath="true" />
       </profile>
     </annotationProcessing>
+    <bytecodeTargetLevel>
```
Comment (smolcoder) [inline .idea/compiler.xml]: why do we need this code to be committed?

File: gradlew [hunk 9]
```diff
@@ -6,12 +6,30 @@
 ##
 ##############################################################################
 
-# Add default JVM options here. You can also use JAVA_OPTS and GRADLE_OPTS to pass JVM options to this script.
-DEFAULT_JVM_OPTS=""
```
Comment (smolcoder) [inline gradlew]: why this file was changed?!

File: genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt (hunk 1 above)
Comment (smolcoder) [inline genequery-rest/src/main/kotlin/gq/rest/services/OverlapService.kt]: It's okay to keep this abstraction

File: .idea/.name (hunk 7 above)
Comment (smolcoder) [inline .idea/.name]: I guess the reason why IDEA is trying to rename the project is your folder's structure.
image
`genequery-kotlin` must be the root folder of the project (even if its name differs from the original), but it is not. Try to make it such.

File: gradlew.bat [hunk 10]
```diff
@@ -8,14 +8,14 @@
 @rem Set local scope for the variables with windows NT shell
```
Comment (smolcoder) [inline gradlew.bat]: You should never touch the gradlew-files (until new version comes) since they're generated automatically.

**Pull request closed without merging**